diff --git a/src/umm_info.c b/src/umm_info.c
index 6868c00..6e25ae3 100644
--- a/src/umm_info.c
+++ b/src/umm_info.c
@@ -20,2 +20,5 @@ UMM_HEAP_INFO ummHeapInfo;
 void *umm_info( void *ptr, int force ) {
+  if(umm_heap == NULL) {
+    umm_init();
+  }
 
diff --git a/src/umm_malloc_cfg.h b/src/umm_malloc_cfg.h
index 6a5a7fc..a74e0a3 100644
--- a/src/umm_malloc_cfg.h
+++ b/src/umm_malloc_cfg.h
@@ -5,6 +5,18 @@
 #ifndef _UMM_MALLOC_CFG_H
 #define _UMM_MALLOC_CFG_H
 
+#include <c_types.h>
+#include <m_printf.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef UMM_MALLOC_H
+#define printf m_printf
+#define puts m_puts
+#endif
+
 /*
  * There are a number of defines you can set at compile time that affect how
  * the memory allocator will operate.
@@ -39,11 +51,11 @@
  * ----------------------------------------------------------------------------
  */
 
-extern char test_umm_heap[];
+extern char _heap_start;
 
 /* Start addresses and the size of the heap */
-#define UMM_MALLOC_CFG_HEAP_ADDR (test_umm_heap)
-#define UMM_MALLOC_CFG_HEAP_SIZE 0x10000
+#define UMM_MALLOC_CFG_HEAP_ADDR ((uint32_t)&_heap_start)
+#define UMM_MALLOC_CFG_HEAP_SIZE ((size_t)(0x3fffc000 - UMM_MALLOC_CFG_HEAP_ADDR))
 
 /* A couple of macros to make packing structures less compiler dependent */
 
@@ -151,11 +163,11 @@ extern char test_umm_heap[];
  * callback is called: `UMM_HEAP_CORRUPTION_CB()`
  */
 
-#define UMM_POISON_CHECK
+// #define UMM_POISON_CHECK
 
 #define UMM_POISON_SIZE_BEFORE 4
 #define UMM_POISON_SIZE_AFTER 4
-#define UMM_POISONED_BLOCK_LEN_TYPE short
+#define UMM_POISONED_BLOCK_LEN_TYPE uint32_t
 
 #ifdef UMM_POISON_CHECK
    void *umm_poison_malloc( size_t size );
@@ -168,4 +180,10 @@ extern char test_umm_heap[];
 #  define POISON_CHECK() 0
 #endif
 
+// #define UMM_HEAP_CORRUPTION_CB() panic()
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* _UMM_MALLOC_CFG_H */
