diff --git a/glue-esp/include-esp/arch/cc.h b/glue-esp/include-esp/arch/cc.h
index 735e700..47413e5 100644
--- a/glue-esp/include-esp/arch/cc.h
+++ b/glue-esp/include-esp/arch/cc.h
@@ -59,6 +59,7 @@ typedef signed     short   s16_t;
 typedef unsigned   long    u32_t;
 typedef signed     long    s32_t;
 typedef unsigned long   mem_ptr_t;
+typedef signed short        sint16_t;
 
 #define S16_F "d"
 #define U16_F "d"
diff --git a/glue-lwip/arch/cc.h b/glue-lwip/arch/cc.h
index 63cd72d..5f100eb 100644
--- a/glue-lwip/arch/cc.h
+++ b/glue-lwip/arch/cc.h
@@ -34,6 +34,8 @@ author: d. gauchard
 
 #include "stdint.h"
 
+typedef signed short        sint16_t;
+
 #ifdef LWIP_BUILD
 
 // define LWIP_BUILD only when building LWIP
