diff --git a/Makefile b/Makefile
index ca234de..08b0fba 100644
--- a/Makefile
+++ b/Makefile
@@ -55,6 +55,12 @@ endif
 ifeq ($(RBOOT_IROM_CHKSUM),1)
 	CFLAGS += -DBOOT_IROM_CHKSUM
 endif
+ifneq ($(RBOOT_ROM1_ADDR),)
+	CFLAGS += -DBOOT_ROM1_ADDR=$(RBOOT_ROM1_ADDR)
+endif
+ifneq ($(RBOOT_ROM2_ADDR),)
+	CFLAGS += -DBOOT_ROM2_ADDR=$(RBOOT_ROM2_ADDR)
+endif
 ifneq ($(RBOOT_EXTRA_INCDIR),)
 	CFLAGS += $(addprefix -I,$(RBOOT_EXTRA_INCDIR))
 endif
diff --git a/rboot.c b/rboot.c
index 6e10a2f..7ac60a8 100644
--- a/rboot.c
+++ b/rboot.c
@@ -250,7 +250,25 @@ static uint8 calc_chksum(uint8 *start, uint8 *end) {
 static uint8 default_config(rboot_config *romconf, uint32 flashsize) {
 	romconf->count = 2;
 	romconf->roms[0] = SECTOR_SIZE * (BOOT_CONFIG_SECTOR + 1);
+
+#ifdef BOOT_ROM1_ADDR
+	romconf->roms[1] = BOOT_ROM1_ADDR;
+#else
 	romconf->roms[1] = (flashsize / 2) + (SECTOR_SIZE * (BOOT_CONFIG_SECTOR + 1));
+#endif
+
+#if defined(BOOT_BIG_FLASH) && defined(BOOT_GPIO_ENABLED)
+	if(flashsize > 0x200000) {
+		romconf->count += 1;
+#ifdef BOOT_ROM2_ADDR
+		romconf->roms[2] = BOOT_ROM2_ADDR;
+#else
+		romconf->roms[2] = 0x310000;
+#endif
+		romconf->gpio_rom = 2;
+	}
+#endif
+
 #ifdef BOOT_GPIO_ENABLED
 	romconf->mode = MODE_GPIO_ROM;
 #endif
