diff -Naur a/Include/pcap/socket.h b/Include/pcap/socket.h
--- a/Include/pcap/socket.h	2019-04-25 17:20:00 +0100
+++ b/Include/pcap/socket.h	2020-06-16 12:20:11 +0100
@@ -49,15 +49,6 @@
   #include <ws2tcpip.h>
 
   /*
-   * Winsock doesn't have this UN*X type; it's used in the UN*X
-   * sockets API.
-   *
-   * XXX - do we need to worry about UN*Xes so old that *they*
-   * don't have it, either?
-   */
-  typedef int socklen_t;
-
-  /*
    * Winsock doesn't have this POSIX type; it's used for the
    * tv_usec value of struct timeval.
    */
