diff --git a/include/flatbuffers/base.h b/include/flatbuffers/base.h
index 54a51aa..2b8bcd3 100644
--- a/include/flatbuffers/base.h
+++ b/include/flatbuffers/base.h
@@ -32,11 +32,11 @@
 #include <cstdlib>
 #include <cstring>
 
-#if defined(ARDUINO) && !defined(ARDUINOSTL_M_H)
-  #include <utility.h>
-#else
+//#if defined(ARDUINO) && !defined(ARDUINOSTL_M_H)
+//  #include <utility.h>
+//#else
   #include <utility>
-#endif
+//#endif
 
 #include <string>
 #include <type_traits>
