diff --git a/BleGamepad.cpp b/BleGamepad.cpp
index 212e6d7..f7fc502 100644
--- a/BleGamepad.cpp
+++ b/BleGamepad.cpp
@@ -834,8 +834,7 @@ void BleGamepad::release(uint8_t b)
 
 uint8_t BleGamepad::specialButtonBitPosition(uint8_t b)
 {
-    if (b >= POSSIBLESPECIALBUTTONS)
-        throw std::invalid_argument("Index out of range");
+    assert(b < POSSIBLESPECIALBUTTONS);
     uint8_t bit = 0;
     for (int i = 0; i < b; i++)
     {
