diff --git a/multipart_parser.c b/multipart_parser.c
index 981dabb..52e0200 100644
--- a/multipart_parser.c
+++ b/multipart_parser.c
@@ -5,9 +5,10 @@
 
 #include "multipart_parser.h"
 
-#include <stdio.h>
-#include <stdarg.h>
 #include <string.h>
+#ifndef tolower
+	int tolower (int __c);
+#endif
 
 static void multipart_log(const char * format, ...)
 {
@@ -43,20 +44,6 @@ do {                                                                   \
 #define LF 10
 #define CR 13
 
-struct multipart_parser {
-  void * data;
-
-  size_t index;
-  size_t boundary_length;
-
-  unsigned char state;
-
-  const multipart_parser_settings* settings;
-
-  char* lookbehind;
-  char multipart_boundary[1];
-};
-
 enum state {
   s_uninitialized = 1,
   s_start,
@@ -77,10 +64,10 @@ enum state {
   s_end
 };
 
-multipart_parser* multipart_parser_init
-    (const char *boundary, const multipart_parser_settings* settings) {
+multipart_parser_t* multipart_parser_init
+    (const char *boundary, const multipart_parser_settings_t * settings) {
 
-  multipart_parser* p = malloc(sizeof(multipart_parser) +
+  multipart_parser_t* p = malloc(sizeof(multipart_parser_t) +
                                strlen(boundary) +
                                strlen(boundary) + 9);
 
@@ -96,22 +83,22 @@ multipart_parser* multipart_parser_init
   return p;
 }
 
-void multipart_parser_free(multipart_parser* p) {
+void multipart_parser_free(multipart_parser_t* p) {
   free(p);
 }
 
-void multipart_parser_set_data(multipart_parser *p, void *data) {
+void multipart_parser_set_data(multipart_parser_t *p, void *data) {
     p->data = data;
 }
 
-void *multipart_parser_get_data(multipart_parser *p) {
+void *multipart_parser_get_data(multipart_parser_t *p) {
     return p->data;
 }
 
-size_t multipart_parser_execute(multipart_parser* p, const char *buf, size_t len) {
+size_t multipart_parser_execute(multipart_parser_t* p, const char *buf, size_t len) {
   size_t i = 0;
   size_t mark = 0;
-  char c, cl;
+  unsigned char c, cl;
   int is_last = 0;
 
   while(i < len) {
diff --git a/multipart_parser.h b/multipart_parser.h
index 05429d7..5fe695e 100644
--- a/multipart_parser.h
+++ b/multipart_parser.h
@@ -11,14 +11,13 @@ extern "C"
 #endif
 
 #include <stdlib.h>
-#include <ctype.h>
 
-typedef struct multipart_parser multipart_parser;
-typedef struct multipart_parser_settings multipart_parser_settings;
+typedef struct multipart_parser multipart_parser_t;
+typedef struct multipart_parser_settings multipart_parser_settings_t;
 typedef struct multipart_parser_state multipart_parser_state;
 
-typedef int (*multipart_data_cb) (multipart_parser*, const char *at, size_t length);
-typedef int (*multipart_notify_cb) (multipart_parser*);
+typedef int (*multipart_data_cb) (multipart_parser_t*, const char *at, size_t length);
+typedef int (*multipart_notify_cb) (multipart_parser_t*);
 
 struct multipart_parser_settings {
   multipart_data_cb on_header_field;
@@ -31,15 +30,28 @@ struct multipart_parser_settings {
   multipart_notify_cb on_body_end;
 };
 
-multipart_parser* multipart_parser_init
-    (const char *boundary, const multipart_parser_settings* settings);
+struct multipart_parser {
+  void* data;
 
-void multipart_parser_free(multipart_parser* p);
+  size_t index;
+  size_t boundary_length;
 
-size_t multipart_parser_execute(multipart_parser* p, const char *buf, size_t len);
+  unsigned char state;
 
-void multipart_parser_set_data(multipart_parser* p, void* data);
-void * multipart_parser_get_data(multipart_parser* p);
+  const multipart_parser_settings_t* settings;
+
+  char* lookbehind;
+  char multipart_boundary[1];
+};
+
+multipart_parser_t* multipart_parser_init(const char* boundary, const multipart_parser_settings_t* settings);
+
+void multipart_parser_free(multipart_parser_t* p);
+
+size_t multipart_parser_execute(multipart_parser_t* p, const char* buf, size_t len);
+
+void multipart_parser_set_data(multipart_parser_t* p, void* data);
+void * multipart_parser_get_data(multipart_parser_t* p);
 
 #ifdef __cplusplus
 } /* extern "C" */
   