diff --git a/src/include/FlashString/ObjectBase.hpp b/src/include/FlashString/ObjectBase.hpp
index 78a3323..b464964 100644
--- a/src/include/FlashString/ObjectBase.hpp
+++ b/src/include/FlashString/ObjectBase.hpp
@@ -43,7 +43,7 @@ public:
 	 */
 	FSTR_INLINE size_t size() const
 	{
-		return ALIGNUP(length());
+		return ALIGNUPW(length());
 	}
 
 	/**
diff --git a/src/include/FlashString/String.hpp b/src/include/FlashString/String.hpp
index 4445a3c..6e8a1f9 100644
--- a/src/include/FlashString/String.hpp
+++ b/src/include/FlashString/String.hpp
@@ -93,7 +93,7 @@ typedef const __FlashStringHelper* flash_string_t;
 #define DEFINE_FSTR_DATA(name, str)                                                                                    \
 	constexpr const struct {                                                                                           \
 		FSTR::ObjectBase object;                                                                                       \
-		char data[ALIGNUP(sizeof(str))];                                                                               \
+		char data[ALIGNUPW(sizeof(str))];                                                                               \
 	} name PROGMEM = {{sizeof(str) - 1}, str};                                                                         \
 	FSTR_CHECK_STRUCT(name);
 
@@ -179,7 +179,7 @@ public:
 	 */
 	size_t size() const
 	{
-		return ALIGNUP(Object::length() + 1);
+		return ALIGNUPW(Object::length() + 1);
 	}
 
 	/**
diff --git a/test/app/string.cpp b/test/app/string.cpp
index 8cd4e81..8641128 100644
--- a/test/app/string.cpp
+++ b/test/app/string.cpp
@@ -103,7 +103,7 @@ public:
 			Serial.print('"');
 			Serial.print(test);
 			Serial.println('"');
-			REQUIRE(sizeof(test) == ALIGNUP(sizeof(DEMO_TEST_TEXT)));
+			REQUIRE(sizeof(test) == ALIGNUPW(sizeof(DEMO_TEST_TEXT)));
 
 			Serial.print("> demoFSTR1 (write): ");
 			Serial.print('"');
@@ -142,7 +142,7 @@ public:
 			Serial.print('"');
 			Serial.write(fsarr, sizeof(fsarr) - 1);
 			Serial.println('"');
-			REQUIRE(sizeof(fsarr) == ALIGNUP(sizeof(DEMO_TEST_TEXT)));
+			REQUIRE(sizeof(fsarr) == ALIGNUPW(sizeof(DEMO_TEST_TEXT)));
 			REQUIRE(memcmp_P(fsarr, DEMO_TEST_TEXT, sizeof(DEMO_TEST_TEXT)) == 0);
 		}
 
