diff --git a/ws_parser.c b/ws_parser.c
index d43adbe..beeb611 100644
--- a/ws_parser.c
+++ b/ws_parser.c
@@ -3,6 +3,7 @@
 #endif
 
 #include "ws_parser.h"
+#include <string.h>
 
 enum {
     S_OPCODE = 0,
@@ -27,6 +28,7 @@ enum {
 void
 ws_parser_init(ws_parser_t* parser)
 {
+    memset(parser, 0, sizeof(ws_parser_t));
     parser->state = S_OPCODE;
     parser->fragment = 0;
 }
@@ -247,7 +249,8 @@ ws_parser_execute(
                 }
 
                 if(parser->mask_flag) {
-                    for(size_t i = 0; i < chunk_length; i++) {
+                    size_t i = 0;
+                    for(i = 0; i < chunk_length; i++) {
                         buff[i] ^= parser->mask[parser->mask_pos++];
                     }
                 }
