diff --git a/ws_parser.c b/ws_parser.c
index 88e4810..e2d1147 100644
--- a/ws_parser.c
+++ b/ws_parser.c
@@ -3,6 +3,7 @@
 #endif
 
 #include "ws_parser.h"
+#include <string.h>
 
 enum {
     S_OPCODE = 0,
@@ -27,6 +28,7 @@ enum {
 void
 ws_parser_init(ws_parser_t* parser, const ws_parser_callbacks_t* callbacks)
 {
+	memset(parser, 0, sizeof(ws_parser_t));
     parser->user_data = NULL;
     parser->callbacks = callbacks;
     parser->state = S_OPCODE;
@@ -244,7 +246,8 @@ ws_parser_execute(ws_parser_t* parser, /* mutates! */ char* buff, size_t len)
                 }
 
                 if(parser->mask_flag) {
-                    for(size_t i = 0; i < chunk_length; i++) {
+                    size_t i = 0;
+                    for(i = 0; i < chunk_length; i++) {
                         buff[i] ^= parser->mask[parser->mask_pos++];
                     }
                 }
